### DISCUSSION TITLE
Connection for S3KeySensor

### DISCUSSION THREAD

**Discussion by M-Waleed-Khan** [apache/airflow#35957] [Q&A]
### Apache Airflow version

Other Airflow 2 version (please specify below)

### What happened

Hi.

I am trying to use S3KeySensor in my dag to trigger a task inside the same dag. here are the configurations I am using for S3KeySensor

from airflow.providers.amazon.aws.sensors.s3 import S3KeySensor
aws_credentials = {
'aws_access_key_id': 'ACCESS_KEY_ID',
'aws_secret_access_key': 'SECRET_ACCESS_KEY'
}

file_sensor_task = S3KeySensor(
task_id='file_sensor',
poke_interval=15, # Check for file every 10 minutes
timeout=60, # Wait up to 30 minutes for the file to appear
bucket_name='BUCKET_NAME',
bucket_key='BUCKET_KEY',
mode='poke',
aws_conn_id=aws_credentials,
dag=DAG_NAME,
)

The reason I am using aws_credentials dictionary is because I am trying to bypass adding the connection details in airflow Connections UI. When I open my connections UI in airflow this is what I get
![image](https://github.com/apache/airflow/assets/48055930/60bda245-5f95-4670-9c08-9a8d9df677e4)

The error I get when I try to use aws_credentials dictionary for aws_conn_id is following:
[2023-11-29, 13:15:49 UTC] {{connection.py:407}} ERROR - Unable to retrieve connection from secrets backend (EnvironmentVariablesBackend). Checking subsequent secrets backend. Traceback (most recent call last): File "/usr/local/airflow/.local/lib/python3.7/site-packages/airflow/models/connection.py", line 400, in get_connection_from_secrets conn = secrets_backend.get_connection(conn_id=conn_id) File "/usr/local/airflow/.local/lib/python3.7/site-packages/airflow/secrets/base_secrets.py", line 64, in get_connection conn_uri = self.get_conn_uri(conn_id=conn_id) File "/usr/local/airflow/.local/lib/python3.7/site-packages/airflow/secrets/environment_variables.py", line 33, in get_conn_uri environment_uri = os.environ.get(CONN_ENV_PREFIX + conn_id.upper()) AttributeError: 'dict' object has no attribute 'upper' [2023-11-29, 13:15:49 UTC] {{connection.py:407}} ERROR - Unable to retrieve connection from secrets backend (MetastoreBackend). Checking subsequent secrets backend. Traceback (most recent call last): File "/usr/local/lib64/python3.7/site-packages/sqlalchemy/engine/base.py", line 1277, in _execute_context cursor, statement, parameters, context File "/usr/local/lib64/python3.7/site-packages/sqlalchemy/engine/default.py", line 608, in do_execute cursor.execute(statement, parameters) psycopg2.ProgrammingError: can't adapt type 'dict'

Please help me by telling me if there is any way I can solve the aws_conn_id error
version 2.2

### What you think should happen instead

successfull connection

### How to reproduce

pass a diction for conn_id

### Operating System

Mac and AWS MWAA both same

### Versions of Apache Airflow Providers

6.0.0

### Deployment

Docker-Compose

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by Taragolis** [comment]
> The error I get when I try to use aws_credentials dictionary for aws_conn_id is following:

`aws_conn_id` - expected only as as string (Connection ID), that is why you get this error, and it can't work with anything else rather than it.

> The reason I am using aws_credentials dictionary is because I am trying to bypass adding the connection details in airflow Connections UI. When I open my connections UI in airflow this is what I get

Difficult to say without error traceback from the webserver.

**Reply by M-Waleed-Khan** [reply]
I am not authorized to create a connection from the airflow UI. is there any other way I can implement a trigger/Sensor on S3 that runs airflow dag when a file is uploaded on s3

**Reply by potiuk** [reply]
You need to ask the one who is authorized to add it. Or get the authorization. Or use any other method you can find in the [Documentation](https://airflow.apache.org/docs/apache-airflow/stable/howto/connection.html)  

Generally if you are asking "CAN I expose sensitive credentials in my dags to others in my DAGand write a code using it" - then yes, of course you CAN do it. You can write your own sensor. This is all python class to implement or even copy the existing Hook to do so, or even set environment variable with those in your DAG and use the existing Hook to do so. 

But whether you SHOULD do that rather than manage the connections the way it is intended, where it is kept outside of your DAG (in your environment or DB or even secret manager) - that's another question

**Reply by M-Waleed-Khan** [reply]
Hi, sorry for delayed response I have been trying some solutions but turns out nothing works for me..

Since @potiuk you mentioned above there is no other way to use S3KeySensor other than adding the connection details in the Airflow UI Connections. I must add the connection details over there. in order to do it I need to access the http://localhost:8080/connection/list/ page which I am unable to do. Here are the traceback from the webser @Taragolis :

```
docker-local-runner-1  | 192.168.65.1 - - [07/Dec/2023:10:32:01 +0000] "POST /task_stats HTTP/1.1" 200 77993 "http://localhost:8080/home" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/116.0.0.0 Safari/537.36"
docker-local-runner-1  | [2023-12-07 10:32:11,144] {{app.py:1892}} ERROR - Exception on /connection/list/ [GET]
docker-local-runner-1  | Traceback (most recent call last):
docker-local-runner-1  |   File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 2447, in wsgi_app
docker-local-runner-1  |     response = self.full_dispatch_request()
docker-local-runner-1  |   File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 1952, in full_dispatch_request
docker-local-runner-1  |     rv = self.handle_user_exception(e)
docker-local-runner-1  |   File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 1821, in handle_user_exception
docker-local-runner-1  |     reraise(exc_type, exc_value, tb)
docker-local-runner-1  |   File "/usr/local/lib/python3.7/site-packages/flask/_compat.py", line 39, in reraise
docker-local-runner-1  |     raise value
docker-local-runner-1  |   File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 1950, in full_dispatch_request
docker-local-runner-1  |     rv = self.dispatch_request()
docker-local-runner-1  |   File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 1936, in dispatch_request
docker-local-runner-1  |     return self.view_functions[rule.endpoint](**req.view_args)
docker-local-runner-1  |   File "/usr/local/airflow/.local/lib/python3.7/site-packages/flask_appbuilder/security/decorators.py", line 148, in wraps
docker-local-runner-1  |     return f(self, *args, **kwargs)
docker-local-runner-1  |   File "/usr/local/airflow/.local/lib/python3.7/site-packages/flask_appbuilder/views.py", line 554, in list
docker-local-runner-1  |     widgets = self._list()
docker-local-runner-1  |   File "/usr/local/airflow/.local/lib/python3.7/site-packages/flask_appbuilder/baseviews.py", line 1134, in _list
docker-local-runner-1  |     page_size=page_size,
docker-local-runner-1  |   File "/usr/local/airflow/.local/lib/python3.7/site-packages/flask_appbuilder/baseviews.py", line 1033, in _get_list_widget
docker-local-runner-1  |     page_size=page_size,
docker-local-runner-1  |   File "/usr/local/airflow/.local/lib/python3.7/site-packages/flask_appbuilder/models/sqla/interface.py", line 471, in query
docker-local-runner-1  |     query_results = query.all()
docker-local-runner-1  |   File "/usr/local/lib64/python3.7/site-packages/sqlalchemy/orm/query.py", line 3373, in all
docker-local-runner-1  |     return list(self)
docker-local-runner-1  |   File "/usr/local/lib64/python3.7/site-packages/sqlalchemy/orm/loading.py", line 100, in instances
docker-local-runner-1  |     cursor.close()
docker-local-runner-1  |   File "/usr/local/lib64/python3.7/site-packages/sqlalchemy/util/langhelpers.py", line 70, in __exit__
docker-local-runner-1  |     with_traceback=exc_tb,
docker-local-runner-1  |   File "/usr/local/lib64/python3.7/site-packages/sqlalchemy/util/compat.py", line 182, in raise_
docker-local-runner-1  |     raise exception
docker-local-runner-1  |   File "/usr/local/lib64/python3.7/site-packages/sqlalchemy/orm/loading.py", line 80, in instances
docker-local-runner-1  |     rows = [proc(row) for row in fetch]
docker-local-runner-1  |   File "/usr/local/lib64/python3.7/site-packages/sqlalchemy/orm/loading.py", line 80, in <listcomp>
docker-local-runner-1  |     rows = [proc(row) for row in fetch]
docker-local-runner-1  |   File "/usr/local/lib64/python3.7/site-packages/sqlalchemy/orm/loading.py", line 601, in _instance
docker-local-runner-1  |     state.manager.dispatch.load(state, context)
docker-local-runner-1  |   File "/usr/local/lib64/python3.7/site-packages/sqlalchemy/event/attr.py", line 322, in __call__
docker-local-runner-1  |     fn(*args, **kw)
docker-local-runner-1  |   File "/usr/local/lib64/python3.7/site-packages/sqlalchemy/orm/mapper.py", line 3397, in _event_on_load
docker-local-runner-1  |     instrumenting_mapper._reconstructor(state.obj())
docker-local-runner-1  |   File "/usr/local/airflow/.local/lib/python3.7/site-packages/airflow/models/connection.py", line 153, in on_db_load
docker-local-runner-1  |     if self.password:
docker-local-runner-1  |   File "/usr/local/lib64/python3.7/site-packages/sqlalchemy/orm/attributes.py", line 365, in __get__
docker-local-runner-1  |     retval = self.descriptor.__get__(instance, owner)
docker-local-runner-1  |   File "/usr/local/airflow/.local/lib/python3.7/site-packages/airflow/models/connection.py", line 238, in get_password
docker-local-runner-1  |     return fernet.decrypt(bytes(self._password, 'utf-8')).decode()
docker-local-runner-1  |   File "/usr/local/lib64/python3.7/site-packages/cryptography/fernet.py", line 194, in decrypt
docker-local-runner-1  |     raise InvalidToken
docker-local-runner-1  | cryptography.fernet.InvalidToken
docker-local-runner-1  | 192.168.65.1 - - [07/Dec/2023:10:32:11 +0000] "GET /connection/list/ HTTP/1.1" 500 1538 "http://localhost:8080/home" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/116.0.0.0 Safari/537.36"


```

**Reply by potiuk** [reply]
Well. You have wrong/invalid fernet token for your installation - this needs to be fixed.
